src/arithmetic/bound_deducer.cc

```diff
@@ -149,7 +150,29 @@ class BoundDeducer: public IRVisitor {
     // always use relax bound
     bool divided = analyzer_.CanProve(result_ % operand == 0);
 
-    result_ = result_ / operand;
+    /*
+     * If both sides are convertible to int constant, then
+     * use floordiv and evaluate the new constant.
+     * This particularly is useful for bound check for conditions on
+     * loop itervars. e.g., an expr like -5/8 should not evaluate to
+     * zero because that indicate that the loop iter var can take on
+     * value of 0 in loop_partition.cc.
+     * TODO: On the other hand, under what conditions floordiv is not
+     * desirable for int constant and you want round towards zero
+     * result?
+     */
+    auto dividend = analyzer_.Simplify(result_);
+    auto divisor = analyzer_.Simplify(operand);
+    auto* dividend_const_value_ptr = as_const_int(dividend);
+    auto* divisor_const_value_ptr = as_const_int(divisor);
+    if (dividend_const_value_ptr && divisor_const_value_ptr) {
+      auto int_bound = analyzer_.const_int_bound(floordiv(dividend, divisor));
+      CHECK_EQ(int_bound->min_value, int_bound->max_value);
+      result_ = Expr(static_cast<int32_t>(int_bound->min_value));
+    }
+    else {
+      result_ = result_ / operand;
+    }
 
     if (!divided) {
```

Comment:
I'm not sure whether the above fix is the right one. Because here it is addressing the non-divisible case and adding compensation. What is the input of this function for the error case you saw? Maybe the compensation path below misses some cases and fixing it there would be better.

Reply:
So the basic idea of loop partitioning is: Original loop has a bunch of if statements for bounds check. Loop partitioning finds those intervals of the loop iter vars for which the conditions, in which that iter var is involved, can be proven to be true or false. So for the interval, given the bounded value of the other var involved in that condition, you can partition the loop with bounds from the interval and remove the conditions. It is this logic, that I updated, finds the interval. In doing so one of the op is divide. Default integer division is round towards zero. So anything between -1 and 0 gets rounded to zero. That is why I added floordiv.
Regarding compensation path missing other cases, it is not clear that it misses other cases but it is possible. Thats why I opened up PR in dmlc/tvm asking for tq's input.